lib/shelley/src/Cardano/Wallet/Shelley/Network.hs

```diff
@@ -0,0 +1,523 @@
+    ByteString
+        -- ^ Concrete representation for bytes string
+    Void
+        -- ^ -- Return type of a network client. Void indicates that the client
+        -- never exits.
+    Void
+        -- ^ Irrelevant for 'InitiatorApplication'. Return type of 'Responder'
+        -- application.
+
+
+--------------------------------------------------------------------------------
+--
+-- Network Client
+
+-- Connect a client to a network, see `newNetworkClient` to construct a network
+-- client interface.
+--
+-- >>> connectClient (newNetworkClient t params) dummyNodeToClientVersion addr
+connectClient
+    :: forall vData. (vData ~ NodeToClientVersionData)
+    => NetworkClient IO
+    -> (vData, CodecCBORTerm Text vData)
+    -> AddrInfo
+    -> IO ()
+connectClient client (vData, vCodec) addr = do
+    let vDict = DictVersion vCodec
+    let versions = simpleSingletonVersions NodeToClientV_1 vData vDict client
+    connectTo (,) versions Nothing addr `catch` handleMuxError
+  where
+    -- `connectTo` might rise an exception: we are the client and the protocols
+    -- specify that only  client can lawfuly close a connection, but the other
+    -- side might just disappear.
+    --
+    -- NOTE: This handler does nothing.
+    handleMuxError :: MuxError -> IO ()
+    handleMuxError = throwIO
+
+-- | A dummy network magic for a local cluster. When connecting to mainnet or
+-- testnet, this should match the underlying network's configuration.
+dummyNodeToClientVersion
+    :: (NodeToClientVersionData, CodecCBORTerm Text NodeToClientVersionData)
+dummyNodeToClientVersion =
+    ( NodeToClientVersionData { networkMagic = 0 }
+    , nodeToClientCodecCBORTerm
+    )
+
+-- | Construct a new network client handling 'ChainSync' and 'LocalTxSubmission'
+-- mini-protocols.
+newNetworkClient
+    :: forall m. (MonadThrow m, MonadST m, MonadTimer m)
+    => Tracer m String
+    -> ChainParameters
+    -> m (NetworkClient m, NetworkLayer m)
+newNetworkClient t params = do
+    chainSyncQueue <- atomically newTQueue
+
+    let client = OuroborosInitiatorApplication $ \pid -> \case
+            ChainSyncWithBlocksPtcl ->
+                chainSyncWithBlocks @m pid t params chainSyncQueue
+            LocalTxSubmissionPtcl ->
+                localTxSubmission @m pid t
+
+    let interface = NetworkLayer
+            { nextBlocks = \p -> ExceptT $ do
+                tvar <- newEmptyTMVarM
+                let request = ReqNextBlocks p (atomically . putTMVar tvar)
+                atomically $ writeTQueue chainSyncQueue request
+                atomically $ takeTMVar tvar
+            , networkTip = do
+                tvar <- newEmptyTMVarM
+                let request = ReqNetworkTip (atomically . putTMVar tvar)
+                atomically $ writeTQueue chainSyncQueue request
+                atomically $ takeTMVar tvar
+            }
+
+    return (client, interface)
+
+-- | Client for the 'Chain Sync' mini-protocol.
```

Comment:
if I understood properly, ouroboros-network is prepared to accommodate any `typed-protocol` which just need to satisfy two things : 1. at any moment only one side of communication is active (ie., about to send) and the other is passive, ie., listening (and being active means `has agency`); 2. protocol has to be described as state transition system, 3. connection have time-ordered guarrantee. If such a typed protocol is delivered then the framework guarrantees no deadlock/race conditions and that state transition will be valid (takes care who needs to send and who needs to listen).

Reply:
Maybe it would be beneficial to put something like this as a comment.

Reply:
> ouroboros-network is prepared to accommodate any typed-protocol

I am not sure about this ... The ouroboros-network does define a few typed-protocols that they call "mini-protocols" (cf section 4. in https://raw.githubusercontent.com/wiki/input-output-hk/cardano-wallet/data_diffusion_and_peer_networking_in_shelley.pdf) 

These protocols are well-defined and we can't really do anything we'd like here. The node will expose and speak a few protocols, and we get to chose amongst them.

Reply:
ok, so if they have already defined mini-protocols then at least they are very well tested (state-machine flavor of testing is straightforward guess here) and we will not need to tests them in our package, right??

Reply:
btw. there is a way to define `type-protocol` I believe : https://github.com/input-output-hk/ouroboros-network/blob/master/typed-protocols/src/Network/TypedProtocol.hs#L84

Reply:
> so if they have already defined mini-protocols then at least they are very well tested

:no_mouth: ...

> and we will not need to tests them in our package

We will test everything we implement in our package! With the integration scenarios, with unit tests and with property-based tests. Now, we won't probably test the internals of the node-to-client protocol, but rather, the usage we make of it :) 

> btw. there is a way to define new type-protocol I believe

Of course! One could use the ouroboros-network as a library and define a new sort of protocol if needed. Yet, a node won't understand _any_ protocol you define just because you've used the ouroboros-network package to implement it. Both peers need to agree on which mini-protocols they know and in which version. There's actually a handshake protocol that does just that: figuring out what protocols are available on both peers.